### DISCUSSION TITLE
Having to use Task.Delay between sends.

### DISCUSSION THREAD

**Discussion by mathew-odwyer** [jchristn/WatsonTcp#292] [Q&A]
Take the following code (light abstracted but the `Send` functions are the same and just serialize an object to bytes to send to my client)

```csharp
        foreach (var remotePlayer in remotePlayers)
        {
            if (remotePlayer.PlayerId == player.PlayerId)
            {
                continue;
            }

            await context.Send(remotePlayer.ConnectionId, new CreateRemotePlayerCommand()
            {
                Player = player,
            }).ConfigureAwait(false);

            await Task.Delay(100).ConfigureAwait(false);

            await context.Send(new CreateRemotePlayerCommand()
            {
                Player = remotePlayer,
            }).ConfigureAwait(false);
        }
```

If I am to remove the `Task.Delay` line only some of the messages are received when more than two or three clients are connected to the server. This is likely a synchronization issue and I was wondering what approach I should take to resolve this?

**Comment by mathew-odwyer** [comment]
Should I:

- Just batch all the `CreateRemotePlayerCommand`s and send it as one command (there could be a decent amount of players leading to a larger amount of data to send).
- Or, should II design some sort of mechanism that will replace the `Task.Delay` and wait for acknowledgement from the client before sending the next `CreateRemotePlayerCommand`?

**Comment by jchristn** [comment]
Hi @softwareantics you shouldn't have to use a `Task.Delay`.  How many concurrent connections do you have in order to reproduce this, and what is the approximate number of payload bytes you are sending per message?  Is there any chance of producing a minimal example that reproduces this using any of the `Test` projects in the repo?

**Reply by mathew-odwyer** [reply]
@jchristn I think that maybe this might be an issue on my client. I am using GML for the client-side and have to handle reading the incoming data. Where can I find the logic in WatsonTcp that handles (from the very beginning) an incoming stream? Best case scenario, I can port the logic that is used to handle incoming "packets" from WatsonTcp over to my client in GML.

**Reply by mathew-odwyer** [reply]
If this doesn't resolve the issue, I'll attempt to setup a scenario where I can reproduce the results using purely C# and WatsonTcp.

**Reply by jchristn** [reply]
Hi @softwareantics I'd recommend looking at [FRAMING.md](https://github.com/dotnet/WatsonTcp/blob/master/FRAMING.md) for an overview.  For the logic, I'd look into the [`WatsonMessage.cs`](https://github.com/dotnet/WatsonTcp/blob/master/src/WatsonTcp/WatsonMessage.cs) class, along with the `DataReceiver` methods in both [server](https://github.com/dotnet/WatsonTcp/blob/49d796e4d794ae1f7e74c5778ae8afb73bb4947b/src/WatsonTcp/WatsonTcpServer.cs#L997) and [client](https://github.com/dotnet/WatsonTcp/blob/49d796e4d794ae1f7e74c5778ae8afb73bb4947b/src/WatsonTcp/WatsonTcpClient.cs#L748).
